### ISSUE TITLE
elasticsearch plugin resource does not allow for JAVA_HOME to be set

### ISSUE THREAD

**Issue by dforste** [elastic/puppet-elasticsearch#619]
There is not a way to install a plugin when using a java install not in the path. 

```
Error: /Stage[main]/Es_strat/Elasticsearch::Plugin[lmenezes/elasticsearch-kopf/v2.1.2]/Elasticsearch_plugin[lmenezes/elasticsearch-kopf/v2.1.2]: Could not evaluate: private method `scan' called for nil:NilClass
Notice: /Stage[main]/Elasticsearch/Elasticsearch::Instance[lvsdevstratsrch01-01]/Exec[mkdir_configdir_elasticsearch_lvsdevstratsrch01-01]: Dependency Elasticsearch_plugin[lmenezes/elasticsearch-kopf/v2.1.2] has failures: true
```

If I run `export JAVA_HOME=/path/java` before puppet the run completes without issue.

If I source the init defaults file like this it seems to work as well: 

```
source /etc/sysconfig/elasticsearch-instance
export JAVA_HOME
```

**Comment by hanej** [comment]
Probably the best way to do this would be to source the /etc/sysconfig/elasticsearch defaults file before installing the plugin.

**Comment by tylerjl** [comment]
Just so I have a better understanding of the use case here, are you setting `JAVA_HOME` in via the `init_defaults` parameter on the elasticsearch class? It should be possible to set the environment variables in the provider when calling the `plugin` command by finding the right file that has the `JAVA_HOME` setting, I just want to be sure we're centralizing that information in the right place (I don't think sourcing a file outright would work in this case since the native provider isn't executing in a shell per se).

**Comment by dforste** [comment]
Any ideas on this moving forward? 
Is there an easy way to simply source the defaults file from this location: 
file { "${elasticsearch::params::defaults_location}/elasticsearch-${name}":
https://github.com/elastic/puppet-elasticsearch/blob/master/manifests/service/init.pp#L120

That would seem like an authoritative way to source the default environment from elasticsearch for the plugin.

**Comment by tylerjl** [comment]
There's not a simple way to source the defaults file because the native provider is [calling the plugin executable from Puppet's executable framework](https://github.com/elastic/puppet-elasticsearch/blob/a98184ec9928ebb3a10178356c7d11136649d663/lib/puppet/provider/elastic_plugin.rb#L129) rather than a shell.

With that being said, as you can see in the ruby file I linked to, the command is wrapped in a block that sets some environment variables, so this is possible, just not with a simple `source $file`.

The way to do this would probably be to 1) pass a parameter to the `elasticsearch_plugin` type for the defaults file, which 2) the provider uses to set the environment before calling `plugin` commands. There's several assumptions here that I'm not taking into account though (i.e. what the effect of pulling in all those new variables into the command might be.)

**Comment by lucabelluccini** [comment]
I think the plugin class now requires the JAVA environment to be setup both for the elasticsearch user AND the puppet environment.

Why don't we add a _JAVA_HOME_ parameter at elasticsearch puppet module level, so:
- it can be reused by those who are not installing Java in a typical path (a workaround for this is to use the init_defaults parameter)
- it avoids to set the JAVA_HOME for the puppet environment

**Comment by tylerjl** [comment]
That sounds doable; though there needs to be a _lot_ of changes to make sure `JAVA_HOME` is used properly everywhere module makes calls the java-based executables (any call to `bin/elasticsearch` [which happen in plugins, when determining the ES version, etc.], setting it for services in defaults files, and so on.)

**Comment by lucabelluccini** [comment]
Do we have any status about that?
I am not a puppet expert, but I am searching for a workaround which doesn't involve the injection of the JAVA_HOME env var through the puppet agent.
I was thinking about implementing a custom elasticsearch_plugin provider which can add the ENV[JAVA_HOME]=...

**Comment by tylerjl** [comment]
I haven't had the opportunity to look at this yet, but feel free to poke around the elasticsearch_plugin provider if you want to have a try at implementing it. The provider already does some environment variable munging for other reasons, so there's some prior examples there.

**Comment by lucabelluccini** [comment]
In the meanwhile, as a workaround, I developed a small code to inject the JAVA_HOME variable.

your_module/lib/puppet/provider/environment_var/environment_var.rb
```
Puppet::Type.type(:environment_var).provide(
  :environment_var,
  :parent => Puppet::Provider
) do
  desc <<-END
    Provides environment vars to the puppet agent
    Example:
    environment_var { 'MYVAR':
      ensure => 'MyValue',
    }
    or
    environment_var { 'MYVAR':
      ensure => absent,
    }
  END

  def ensure
    # To avoid change notifications, we set the value at check time
    if resource[:ensure] == :absent
      ENV.delete(resource[:name])
    else
      ENV[resource[:name]] = resource[:ensure]
    end
    resource[:ensure]
  end

  def ensure=(new_value)
    # Nothing to do, as value is always set in the ensure
    # Doing so, no change will be reported
    # This method will actually never been called
    # Since "ensure" returns the same value being requested
  end

end
```

your_module/lib/puppet/type/environment_var.rb
```
Puppet::Type.newtype(:environment_var) do

  @doc = "Environment variable type"

  newparam(:name, :namevar => true) {
    desc "The name of the environment variable."
  }

  newproperty(:ensure) do
    newvalues :absent
    newvalues /[a-zA-Z_]+[a-zA-Z0-9_]*/
  end

end
```

And can be used via
```
environment_var { 'JAVA_HOME':
      ensure   => 'yourpath to java',
    }
```

**Comment by nixgadget** [comment]
For anyone having problems of passing in `JAVA_HOME` I did the following,

```
elasticsearch::instance { "es-test":
    ensure => present,
    init_defaults => {
        'JAVA_HOME' => "\"${jre_path}\""
    },
    .....
}
```

in this case `jre_path` is something like `/usr/local/jre1.8` or similar

**Comment by tylerjl** [comment]
Support for this has been merged in from #902
